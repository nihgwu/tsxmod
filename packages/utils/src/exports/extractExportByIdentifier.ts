import type { SourceFile } from 'ts-morph'
import { Node, ts } from 'ts-morph'

/** Extract a single export and its local dependencies from a source file. */
export function extractExportByIdentifier(
  sourceFile: SourceFile,
  identifier: string
) {
  /** Remove named exports: export { useHover } from 'hooks' */
  sourceFile.getExportDeclarations().forEach((declaration) => {
    declaration.remove()
  })

  /** Collect remaining exports and remove any declarations that don't have references. */
  sourceFile.getExportedDeclarations().forEach((declarations) => {
    declarations.forEach((declaration) => {
      if (Node.isSourceFile(declaration) || Node.isExpression(declaration)) {
        return
      }

      const exportIdentifier = declaration.getFirstDescendantByKind(
        ts.SyntaxKind.Identifier
      )!

      if (exportIdentifier.getText() !== identifier) {
        // TODO: we need a findReferencesInSourceFile() helper
        const references = exportIdentifier.findReferences()

        if (references.length <= 1) {
          declaration.remove()
        }
      }
    })
  })

  /** Finally, fix missing references until we have an equal result. */
  let lastFullText

  while (lastFullText !== sourceFile.getFullText()) {
    lastFullText = sourceFile.getFullText()
    sourceFile.fixUnusedIdentifiers()
  }

  return lastFullText.trim()
}
