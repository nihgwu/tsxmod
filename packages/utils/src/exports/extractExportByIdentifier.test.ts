import { readFileSync } from 'fs'
import { Project } from 'ts-morph'
import { extractExportByIdentifier } from './extractExportByIdentifier'

const kitchenSinkSource = readFileSync('fixtures/kitchen-sink.tsx', 'utf8')

test('extracts export and dependencies', () => {
  const project = new Project({ useInMemoryFileSystem: true })
  const sourceFile = project.createSourceFile('test.tsx', kitchenSinkSource)
  const codeString = extractExportByIdentifier(sourceFile, 'Box')

  expect(codeString).toMatchSnapshot()
})
