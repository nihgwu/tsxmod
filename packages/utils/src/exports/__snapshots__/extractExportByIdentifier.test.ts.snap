// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extracts export and dependencies 1`] = `
"type SystemProps = { as: any }

type BoxProps = { children: any } & SystemProps

export const Box = (props: BoxProps) => <div {...props} />"
`;
